File: NeetCode/LinkedList/Easy/234.PalindromeLinkedList.py
# Given the head of a singly linked list, return true if it is a 
# palindrome or false otherwise.

class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def isPalindrome(head):
    slow, fast = head, head

    while fast and fast.next:
        fast = fast.next.next
        slow = slow.next

    previous_node = None
    while slow:
        next_node = slow.next
        slow.next = previous_node
        previous_node = slow
        slow = next_node

    left, right = head, previous_node
    while right:
        if left.val != right.val:
            return False
        
        left = left.next
        right = right.next

    return True

File: swift/Neetcode/LinkedList/Easy/234.PalindromeLinkedList.swift
// Given the head of a singly linked list, return true if it is a 
// palindrome or false otherwise.

// End node shared both but while logic checks it for even/odd
func isPalindrome(_ head: ListNode?) -> Bool {
    var slow = head, fast = head

    while let next = fast?.next {
        slow = slow?.next
        fast = next.next
    }

    var previous: ListNode? = nil
    var current = slow 

    while current != nil {
        var next = current?.next
        current?.next = previous
        previous = current
        current = next
    }

    var left = head, right = previous

    while let node1 = left, let node2 = right {
        if node1.val != node2.val { return false }

        left = node1.next
        right = node2.next
    }

    return true
}